src/qibolab/instruments/zhinst.py

```diff
@@ -872,32 +872,53 @@ def play_sweep(self, exp, qubit, pulse, section):
                     exp, qubit, pulse, section, parameters, partial_sweep
                 )
 
-    def couplerflux(self, exp, couplers):
-        """Coupler flux for bias sweep or pulses."""
+    def couplerflux(self, exp: lo.Experiment, couplers: Dict[str, Coupler]):
+        """Coupler flux for bias sweep or pulses.
+
+        Args:
+            exp (lo.Experiment): laboneq experiment on which register sequences.
+            couplers (dict[str, Coupler]): coupler on which pulses are played.
+        """
         for coupler in couplers.values():
             c = coupler.name  # pylint: disable=C0103
-            with exp.section(uid=f"sequence_bias_coupler{c}"):
-                i = 0
-                time = 0
-                for pulse in self.sequence[f"couplerflux{c}"]:
-                    pulse.zhpulse.uid += str(i)
-                    exp.delay(
-                        signal=f"couplerflux{c}",
-                        time=round(pulse.pulse.start * NANO_TO_SECONDS, 9) - time,
-                    )
-                    time = round(pulse.pulse.duration * NANO_TO_SECONDS, 9) + round(
-                        pulse.pulse.start * NANO_TO_SECONDS, 9
-                    )
-                    if isinstance(pulse, ZhSweeperLine):
-                        self.play_sweep(exp, coupler, pulse, section="couplerflux")
-                    elif isinstance(pulse, ZhSweeper):
-                        self.play_sweep(exp, coupler, pulse, section="couplerflux")
-                    elif isinstance(pulse, ZhPulse):
-                        exp.play(signal=f"couplerflux{c}", pulse=pulse.zhpulse)
-                    i += 1
-
-    def flux(self, exp, qubits):
-        """Qubit flux for bias sweep or pulses."""
+            time = 0
+            i = 0
+            if len(self.sequence[f"couplerflux{c}"]) != 0:
```

Comment:
Do I understand correctly, that you have this check here, so that `self.sub_sequences[f"couplerflux{c}"]` in the next loop does not throw `KeyError` in case the sequence is empty?

Reply:
In principle, if `self.sequence[X]` has zero length, `self.sub_sequences[X]` is a list of empty lists, therefore the `for pulse in sequence` loop is just skipped. The same applies also for `drive` and `flux` which are (at least for the parts I changed) just a copy-paste, as you noticed.
My only point is that if the length is zero, there might be no `ExperimentSignal` associated to that sequence (for sure for `drive`). Of course we can change lines 718-734 accordingly.

Reply:
> My only point is that if the length is zero, there might be no ExperimentSignal associated to that sequence

Which will mean that `self.sub_sequences[X]` will raise an exception, because a key called `X` does not exist in the `self.sub_sequences` dict, right? I think to avoid this you can take a simpler approach - instead of having the if statement to guard the entire code, you could remove this if overall and replace `self.sub_sequences[f"couplerflux{c}"]` with `self.sub_sequences.get(f"couplerflux{c}", [])`

Reply:
oks, then let's just remove the if